Pyomo.DoE: adding more verbose output for sensitivity analysis

smondal13 opened the pull request:
<!-- ##################################################################### -->
<!-- PLEASE READ BEFORE OPENING THIS PULL REQUEST -->

<!-- All changes must adhere to PEP8 standards as enforced by Black. -->
<!-- If your changes do NOT adhere, the test suite will fail. -->
<!-- Please read our Contributing guide for instructions on how to apply these standards. -->
<!-- Contributing Guide: https://pyomo.readthedocs.io/en/stable/contribution_guide.html -->
<!-- ##################################################################### -->

<!-- DO NOT DELETE OR IGNORE THIS TEMPLATE. Failing to adhere to this template and provide the necessary information may lead to your Pull Request being closed without consideration. -->

## Fixes # .

## Summary/Motivation:
- Adding more verbose is helpful to analyze the problem
- Adding a new method to check the FIM in the utils.py can help in the general usage of the method.
- Adding a new standalone function in utils.py to calculate different metrics of the FIM is helpful in case the user does not want to build the `get_labeled_model()`.
- Adding arguments in the `reactor_example.py` will help with the testing.

## Changes proposed in this PR:
- added eigenvalues, determinant of FIM, and trace of FIM for sensitivity output in `compute_FIM_full_factorial()`
- added `check_FIM()` in utils.py by separating it from `check_model_FIM()` in doe.py
- added `compute_FIM_metrics()` in utils.py
- added `get_FIM_metrics()` in utils.py
- added arguments in the `reactor_example.py`

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

smondal13 commented:
@adowling2 @djlaky ready for early feedback

adowling2 commented:
Next step:
- [ ] Also add solver status

jsiirola commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1521,17 +1525,22 @@ def compute_FIM_full_factorial(
+            E_ind = 0 # np.argmin(E_vals.real)  # Grab index of minima to check imaginary
             # Warn the user if there is a ``large`` imaginary component (should not be)
             if abs(E_vals.imag[E_ind]) > 1e-8:
                 self.logger.warning(
-                    "Eigenvalue has imaginary component greater than 1e-6, contact developers if this issue persists."
+                    "Eigenvalue has imaginary component greater than 1e-8, contact developers if this issue persists."
```
nit: this is why "embedded constants are evil".  An improved solution would be to declare a module-level constant (maybe `IMAG_THRESHOLD`) and then use that both in the test and in the message.

adowling2 commented:
@smondal13 See https://github.com/Pyomo/pyomo/pull/3532 for an example of what @jsiirola is suggesting

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1374,6 +1381,31 @@ def check_model_FIM(self, model=None, FIM=None):
+            "FIM provided matches expected dimensions from model and is approximately positive (semi) definite."
+        )
+
+    # TODO: Make this a static method
```
You can remove this TODO.

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1575,15 +1607,26 @@ def compute_FIM_full_factorial(
+                _compute_FIM_metrics(FIM)
+            )
+            """
+            Since we are using the `_compute_FIM_metrics()` to compute the metrics in this chunk of code., This chunk of code is ready to be deleted.
```
Please delete if you feel it is no longer needed. It is in the Git history in case we need to find it.

adowling2 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -2314,3 +2362,95 @@ def _sgn(self, p):
+    E_vals, E_vecs = np.linalg.eig(FIM)
+    E_ind = np.argmin(E_vals.real)  # index of smallest eigenvalue
+
+    # Warn the user if there is a ``large`` imaginary component (should not be)
```
Will you ever get to this warning? Or will _check_FIM throw an error first?

smondal13 replied:
@adowling2 `_check_FIM()`, will not trigger this warning. `_check_FIM()` only checks whether the FIM is square, symmetric, and positive definite. Although this method checks for the symmetry of the FIM with a tolerance, and it is unlikely that the FIM will have large imaginary eigenvalues; we have kept the warning there just in case there is a large imaginary eigenvalue.

adowling2 replied:
@smondal13 Thank you.

@blnicho @mrmundt Is there a best practice for testing this warning message is properly thrown?

adowling2 reviewed (commented):
@smondal13 This looks really close. I left a few minor comments.

@blnicho @jsiirola @mrmundt This is ready for a design review from the Pyomo team. I am moving it on the project board.

adowling2 commented:
Also, @smondal13, please remove the draft status.

adowling2 reviewed (approved):
Pending input from the Pyomo team, I think this PR is ready to merge.

adowling2 commented:
@smondal13 Tag folks when this is ready for review

smondal13 commented:
@blnicho @mrmundt This PR is ready for review.

blnicho commented:
@smondal13 could you run `black` on your code? The linter is not happy with a couple of your import statements.

See: https://pyomo.readthedocs.io/en/latest/contribution_guide.html#coding-standards

blnicho commented on pyomo/contrib/doe/tests/test_doe_FIM_metrics.py:
```diff
@@ -0,0 +1,73 @@
+            _compute_FIM_metrics(FIM)
+        )
+
+        # expected results
+        det_expected = np.linalg.det(FIM)
+        D_opt_expected = np.log10(det_expected)
+
+        trace_expected = np.trace(FIM)
+        A_opt_expected = np.log10(trace_expected)
+
+        E_vals_expected, E_vecs_expected = np.linalg.eig(FIM)
+        min_eigval = np.min(E_vals_expected.real)
+        if min_eigval <= 0:
+            E_opt_expected = np.nan
+        else:
+            E_opt_expected = np.log10(min_eigval)
+
+        cond_expected = np.linalg.cond(FIM)
+
+        ME_opt_expected = np.log10(cond_expected)
```
You typically want to hard-code the expected results in the test instead of copying the function code here. This isn't really testing anything except that the code is the same in both places. If the numpy behavior ever changed resulting in different values, this test would not catch it and would always pass.

blnicho commented on pyomo/contrib/doe/tests/test_doe_FIM_metrics.py:
```diff
@@ -0,0 +1,73 @@
+
+
+class TestFIMWarning(unittest.TestCase):
+    def test_FIM_eigenvalue_warning(self):
```
I think this looks good.

blnicho commented on pyomo/contrib/doe/tests/test_doe_FIM_metrics.py:
```diff
@@ -0,0 +1,73 @@
+        self.assertEqual(ME_opt, ME_opt_expected)
+
+
+class TestFIMWarning(unittest.TestCase):
```
You should include this as another test under the existing class. No need to have separate classes.

```suggestion
```

smondal13 replied:
Done.

blnicho commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1383,6 +1390,30 @@ def check_model_FIM(self, model=None, FIM=None):
+            "FIM provided matches expected dimensions from model and is approximately positive (semi) definite."
+        )
+
+    @staticmethod
+    def _check_FIM(FIM):
```
Why did you add this as a `staticmethod` instead of a utility function outside of the class?

smondal13 replied:
All the standalone functions are moved to utils.py

blnicho commented on pyomo/contrib/doe/doe.py:
```diff
@@ -2323,3 +2358,95 @@ def _sgn(self, p):
             return -1
+
+
+# loggers for the functions
+function_logger = logging.getLogger(__name__)
+function_logger.setLevel(logging.WARNING)
```
The logger for this module is retrieved above, with the logging level set based on the user input. See:
https://github.com/Pyomo/pyomo/blob/9efc82d3e18777a2c725586c036e209f6cbb79e9/pyomo/contrib/doe/doe.py#L218-L220
I'm worried that grabbing the logger for this module again and setting the level here will lead to weird edge cases with the logging code in the class. What's the reasoning for manually setting the level here?

@adowling2 would you be opposed to moving these stand-alone functions into `pyomo.contrib.doe.utils`?

blnicho replied:
If you move these functions into `pyomo.contrib.doe.utils`, then I would also suggest combining the test files in this PR into a single `test_utils.py` file.

smondal13 commented on pyomo/contrib/doe/doe.py:
```diff
@@ -1479,7 +1506,22 @@ def compute_FIM_full_factorial(
+            "eigval_max": list of maximum eigenvalues
+            "det_FIM": list of determinants
+            "trace_FIM": list of traces
+            "solve_time": list of solve times
```
I have run black, and the file was not changed. I have changed  some formatted string manually which black did not catch. Hope this solves this issue

Also, I have reformatted. I think now it should format nicely.
https://github.com/smondal13/pyomo/blob/6d0ab44e7896b2b3290c5e4b4d0e1c0b02434426/pyomo/contrib/doe/utils.py#L228-L237

smondal13 commented:
@blnicho I have implemented your suggestions. It is ready for review now.

blnicho commented:
> My memory (which may be incorrect) is that @smondal13 ran black without the correct inputs and it changed the string formatting from single quote to double quote. I see this change around line 106 in test_doe_solve.py and a few other places.
> 
> Does this matter or is it okay? @blnicho @jsiirola @mrmundt

It only touches a handful of files in Pyomo.DoE so I think we decided it's fine to leave it. Just make sure to use the `black` options we talked about for future development.

blnicho commented on pyomo/contrib/doe/tests/test_doe_solve.py:
```diff
@@ -11,6 +11,7 @@
 import json
 import logging
 import os.path
+import idaes.core.solvers.get_solver
```
What is this import being used for? We generally want to avoid importing anything from IDAES in Pyomo since it creates a circular dependency.

adowling2 replied:
@smondal13 In our group manual, I added instructions on making the "good Ipopt" executable accessible for all condo environments in macOS. I also posted this in our group Slack a few weeks ago. Going this will mean you do not need IDAES for the ipopt solver.

smondal13 commented:
@blnicho, I have implemented the changes you suggested, and it is ready for review.

blnicho commented:
@smondal13 here are the tests that are failing:

```
=========================== short test summary info ============================
FAILED pyomo/contrib/doe/tests/test_doe_errors.py::TestReactorExampleErrors::test_reactor_check_bad_prior_not_symmetric - ImportError: cannot import name '_SMALL_TOLERANCE_SYMMETRY' from 'pyomo.contrib.doe.doe' (/home/runner/work/pyomo/pyomo/pyomo/contrib/doe/doe.py)
FAILED pyomo/contrib/doe/tests/test_doe_solve.py::TestDoe::test_doe_full_factorial - AssertionError: False is not true
```

The first one looks like an import that needs to be changed in `test_doe_errors.py` from moving the tolerance constants into `utils.py`. The second one looks like some values are not the same when checking "log10 D-opt" across operating systems. You may have to modify the test to print out the values to see how different they are. It's possible it can be fixed by adjusting the tolerance of the `np.allclose` check.

smondal13 commented:
@blnicho, I have added the print statement, and on my Mac, the values of "log10 D-opt" are exactly the same as the "log10_D_opt_expected" values. `test_doe_solve.py`  passed all the tests on my machine. As per your suggestion, I have changed atol = 1e-6 from its default value, atol=1e-8.  I hope this works for other operating systems. 

I have also solved the import issue.

smondal13 commented on pyomo/contrib/doe/tests/test_doe_solve.py:
```diff
@@ -443,5 +445,84 @@ def test_reactor_grid_search_bad_model(self):
+
+        self.assertTrue(
+            np.allclose(ff_results["log10 D-opt"], log10_D_opt_expected, atol=1e-6)
+        )
```
@blnicho, I have changed the design_ranges in for the example in `test_doe_solve.py` . Previously, the values of D-opt were very small. This change will probably solve the numerical issue.

blnicho commented on pyomo/contrib/doe/tests/test_doe_solve.py:
```diff
@@ -443,5 +445,102 @@ def test_reactor_grid_search_bad_model(self):
         )
 
 
```
```suggestion

@unittest.skipIf(not ipopt_available, "The 'ipopt' solver is not available")
```
@smondal13 I think you just need to skip these tests when ipopt is not available to get the last check passing.
